Импортируем pandas для анализа данных

In [1]:
import pandas as pd

Импортируем три датасета

In [2]:
ads = pd.read_csv('data/ads.csv')
leads = pd.read_csv('data/leads.csv')
purchases = pd.read_csv('data/purchases.csv')

Отображает по 1 строке каждого датафрейма, чтобы понимать, что они из себя представляют

In [3]:
ads.head(1)

created_at   d_ad_account_id d_utm_source d_utm_medium  d_utm_campaign  \
0  2022-08-08  xo-for-client-ya       yandex          cpc        48306435   

   d_utm_content  d_utm_term  m_clicks  m_cost  
0     8404700756         NaN       2.0   48.84

In [4]:
leads.head(1)

lead_created_at                               lead_id d_lead_utm_source  \
0      2022-10-16  d8c4a9e8-4d4a-11ed-907e-848f69e142cf               NaN   

  d_lead_utm_medium d_lead_utm_campaign d_lead_utm_content d_lead_utm_term  \
0               NaN                 NaN                NaN             NaN   

                              client_id  
0  d8c4a9e9-4d4a-11ed-907e-848f69e142cf

In [5]:
purchases.head(1)

purchase_created_at                           purchase_id  \
0          2022-02-21  d3198d39-6d16-40c9-bff7-aa28bd6e2991   

                              client_id  m_purchase_amount  
0  7011bdcd-6fd8-11e7-80fc-c412f533dba1             9950.0

А также посмотрим на кол-во строк/столбцов

In [6]:
print(ads.shape)
print(leads.shape)
print(purchases.shape)

(8602, 9)
(23540, 8)
(66180, 4)


Проверяем типы данных наших датафреймов

In [7]:
ads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8602 entries, 0 to 8601
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   created_at       8602 non-null   object 
 1   d_ad_account_id  8602 non-null   object 
 2   d_utm_source     8602 non-null   object 
 3   d_utm_medium     8602 non-null   object 
 4   d_utm_campaign   8602 non-null   int64  
 5   d_utm_content    8602 non-null   int64  
 6   d_utm_term       0 non-null      float64
 7   m_clicks         8602 non-null   float64
 8   m_cost           8602 non-null   float64
dtypes: float64(3), int64(2), object(4)
memory usage: 605.0+ KB


In [8]:
leads.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23540 entries, 0 to 23539
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   lead_created_at      23540 non-null  object
 1   lead_id              23540 non-null  object
 2   d_lead_utm_source    12331 non-null  object
 3   d_lead_utm_medium    10100 non-null  object
 4   d_lead_utm_campaign  9569 non-null   object
 5   d_lead_utm_content   3338 non-null   object
 6   d_lead_utm_term      328 non-null    object
 7   client_id            18601 non-null  object
dtypes: object(8)
memory usage: 1.4+ MB


In [9]:
purchases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66180 entries, 0 to 66179
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   purchase_created_at  66180 non-null  object 
 1   purchase_id          66180 non-null  object 
 2   client_id            66179 non-null  object 
 3   m_purchase_amount    66180 non-null  float64
dtypes: float64(1), object(3)
memory usage: 2.0+ MB


В датафреймах присутствуют переменные с датами, которые желательно переконвертировать в тип данных datetime

In [10]:
ads['created_at'] = pd.to_datetime(ads['created_at'])
leads['lead_created_at'] = pd.to_datetime(leads['lead_created_at'])
purchases['purchase_created_at'] = pd.to_datetime(purchases['purchase_created_at'])

In [11]:
print(ads['created_at'].info())
print(leads['lead_created_at'].info())
print(purchases['purchase_created_at'].info())

<class 'pandas.core.series.Series'>
RangeIndex: 8602 entries, 0 to 8601
Series name: created_at
Non-Null Count  Dtype         
--------------  -----         
8602 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 67.3 KB
None
<class 'pandas.core.series.Series'>
RangeIndex: 23540 entries, 0 to 23539
Series name: lead_created_at
Non-Null Count  Dtype         
--------------  -----         
23540 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 184.0 KB
None
<class 'pandas.core.series.Series'>
RangeIndex: 66180 entries, 0 to 66179
Series name: purchase_created_at
Non-Null Count  Dtype         
--------------  -----         
66180 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 517.2 KB
None


В ads столбцы d_utm_campaign и d_utm_content в формате int. Переконвертируем их в str

In [12]:
ads['d_utm_campaign'] = ads['d_utm_campaign'].astype('str')
ads['d_utm_content'] = ads['d_utm_content'].astype('str')

Проверяем результат

In [13]:
ads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8602 entries, 0 to 8601
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   created_at       8602 non-null   datetime64[ns]
 1   d_ad_account_id  8602 non-null   object        
 2   d_utm_source     8602 non-null   object        
 3   d_utm_medium     8602 non-null   object        
 4   d_utm_campaign   8602 non-null   object        
 5   d_utm_content    8602 non-null   object        
 6   d_utm_term       0 non-null      float64       
 7   m_clicks         8602 non-null   float64       
 8   m_cost           8602 non-null   float64       
dtypes: datetime64[ns](1), float64(3), object(5)
memory usage: 605.0+ KB


В датафрейме ads столбец d_utm_term состоит исключительно из NaN. Его можно удалить

In [14]:
ads.drop('d_utm_term', axis=1, inplace=True)

In [15]:
ads.columns

Index(['created_at', 'd_ad_account_id', 'd_utm_source', 'd_utm_medium',
       'd_utm_campaign', 'd_utm_content', 'm_clicks', 'm_cost'],
      dtype='object')

Описательная статистика наших датафреймов

In [16]:
ads.describe()

created_at     m_clicks       m_cost
count                           8602  8602.000000  8602.000000
mean   2022-05-17 17:09:41.725180160     2.185189    52.446061
min              2022-01-03 00:00:00     0.000000     0.000000
25%              2022-03-30 00:00:00     0.000000     0.000000
50%              2022-05-12 12:00:00     0.000000     0.000000
75%              2022-07-02 00:00:00     3.000000    60.147000
max              2022-09-14 00:00:00    39.000000  1103.616000
std                              NaN     3.945861   102.918082

In [17]:
leads.describe()

lead_created_at
count                          23540
mean   2022-06-02 16:26:46.389125120
min              2022-01-01 00:00:00
25%              2022-03-24 00:00:00
50%              2022-06-09 00:00:00
75%              2022-08-15 00:00:00
max              2022-10-16 00:00:00

In [18]:
purchases.describe()

purchase_created_at  m_purchase_amount
count                          66180       66180.000000
mean   2022-05-27 03:59:52.166817792        2279.621487
min              2022-01-01 00:00:00           0.000000
25%              2022-03-20 00:00:00          60.000000
50%              2022-05-27 00:00:00         210.000000
75%              2022-08-04 00:00:00        1800.000000
max              2022-10-16 00:00:00      260000.000000
std                              NaN        4518.129716

**ads**: имеются данные о 8602 рекламных компаниях (возможно есть дубликаты, это еще предстоит выяснить), рекламные компании были созданы за период с января по сентябрь 2022 года<br>
**leads**: лидов создано 23540 (также возможно есть дубликаты). Лидов(заявок) почти в три раза больше чем рекламных компаний, что логично. Период с января по октябрь 2022 года<br>
**purchases**: данные о 66180 покупках (проверю на наличие дубликатов). Даты покупок совпадают полностью с датами лидов. Медианная стоимость покупки сильно разнится с со средней стоимостью. Это свидетельствует о наличии выбросов. Похоже, что так оно и есть, учитывая, что максимальная стоимость покупки 260000, что кажется не правдоподобным для салона красоты, но заниматься выбросами не буду, оставлю, как есть. В реальных условиях нужно было бы обратиться к эксперту по предметной области за консультацией.

Создам дополнительные столбцы с датами в формате год-месяц. Они понадобятся для группировки

In [19]:
ads['ads_year_month'] = ads['created_at'].dt.strftime('%Y-%m')
leads['leads_year_month'] = leads['lead_created_at'].dt.strftime('%Y-%m')
purchases['purchases_year_month'] = purchases['purchase_created_at'].dt.strftime('%Y-%m')

In [20]:
print(ads['ads_year_month'].head(1))
print(leads['leads_year_month'].head(1))
print(purchases['purchases_year_month'].head(1))

0    2022-08
Name: ads_year_month, dtype: object
0    2022-10
Name: leads_year_month, dtype: object
0    2022-02
Name: purchases_year_month, dtype: object


Для финальной таблички за основу возьмем датафрейм ads. На основе его меток и их содержимого будем джойнить с датафреймом leads.

In [21]:
ads_clicks_cost = ads.groupby(['ads_year_month', 'd_utm_source', 'd_utm_medium', 'd_utm_campaign']).agg({'m_clicks': 'sum', 'm_cost': 'sum'}).reset_index().rename(columns={'m_clicks': 'clicks', 'm_cost': 'campaign_cost'})
ads_clicks_cost.head()

ads_year_month d_utm_source d_utm_medium d_utm_campaign  clicks  \
0        2022-01       yandex          cpc       48306435    31.0   
1        2022-01       yandex          cpc       48306450   956.0   
2        2022-02       yandex          cpc       48306435    46.0   
3        2022-02       yandex          cpc       48306450   809.0   
4        2022-03       yandex          cpc       48306435   367.0   

   campaign_cost  
0        424.608  
1      17355.324  
2        541.716  
3      16806.492  
4       7587.384

Проверим датафреймы на наличие дубликатов

In [22]:
print(ads.duplicated().sum())
print(leads.duplicated().sum())
print(purchases.duplicated().sum())

0
0
0


Датафреймы ads и leads будем джойнить по следующим столбцам из ads(и аналогичным им из leads): created_at, d_utm_source, d_utm_medium, d_utm_campaign, d_utm_content<br>
Важно чтобы в обоих датафреймах значения этих столбцов совпадали

In [23]:
print(ads['d_utm_source'].unique())
print(ads['d_utm_medium'].unique())
print(ads['d_utm_campaign'].nunique())
print(ads['d_utm_content'].nunique())

['yandex']
['cpc']
10
121


Перед тем соединить ads и leads, проведу чистку leads на предмет значений, отсутствующих в метках ads, а также NaN

In [24]:
leads_cl = leads[(leads['d_lead_utm_source'] == 'yandex') & (leads['d_lead_utm_medium'] == 'cpc')]
leads_cl = leads_cl[leads_cl['d_lead_utm_campaign'].isin(ads['d_utm_campaign'].unique())]
leads_cl = leads_cl[leads_cl['d_lead_utm_campaign'].isin(ads['d_utm_campaign'].unique())]
leads_cl = leads_cl[leads_cl['d_lead_utm_content'].isin(ads['d_utm_content'].unique())]

In [25]:
leads_cl.isnull().sum()

lead_created_at           0
lead_id                   0
d_lead_utm_source         0
d_lead_utm_medium         0
d_lead_utm_campaign       0
d_lead_utm_content        0
d_lead_utm_term        1677
client_id               387
leads_year_month          0
dtype: int64

Удаляем столбец d_lead_utm_term из leads (в ads уже удалили - он был пустой). Также дропаем строки со значением NaN из client_id т.к. с отсутствующими данными не удастся сджойнить датафрейм с purchases по client_id

In [26]:
leads_cl.drop('d_lead_utm_term', axis=1, inplace=True)
leads_cl.dropna(subset=['client_id'], axis='rows', inplace=True)

In [27]:
leads_cl.isnull().sum()

lead_created_at        0
lead_id                0
d_lead_utm_source      0
d_lead_utm_medium      0
d_lead_utm_campaign    0
d_lead_utm_content     0
client_id              0
leads_year_month       0
dtype: int64

Джойним датафремы ads и leads

In [28]:
ads_leads = pd.merge(ads, leads_cl, how='inner', left_on=['created_at','d_utm_source', 'd_utm_medium', 'd_utm_campaign', 'd_utm_content'], right_on=['lead_created_at', 'd_lead_utm_source', 'd_lead_utm_medium', 'd_lead_utm_campaign', 'd_lead_utm_content'])
ads_leads.head()

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
0 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
1 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
2 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
3 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
4 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   

  d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  \
0    8404701659      15.0  496.896        2022-08      2022-08-08   
1    8404701661       5.0  124.704        2022-08      2022-08-08   
2    8404701665      20.0  385.896        2022-08      2022-08-08   
3    8404701665      20.0  385.896        2022-08      2022-08-08   
4    8404701665      20.0  385.896        2022-08      2022-08-08   

                                lead_id d_lead_utm_source d_lead_utm_medium  \
0  136c5ccc-16e9-11ed-9074-848f69e142cf            yandex               cpc   
1  8c5a5080-16f8-11ed-9074-848f69e142cf            yandex               cpc   
2  dd50a605-1714-11ed-9074-848f69e142cf            yandex               cpc   
3  7f1ce442-1707-11ed-9074-848f69e142cf            yandex               cpc   
4  434869a7-1704-11ed-9074-848f69e142cf            yandex               cpc   

  d_lead_utm_campaign d_lead_utm_content  \
0            48306450         8404701659   
1            48306450         8404701661   
2            48306450         8404701665   
3            48306450         8404701665   
4            48306450         8404701665   

                              client_id leads_year_month  
0  6ba575d8-f458-11ec-9070-848f69e142cf          2022-08  
1  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08  
2  ca65c322-1ac9-11ec-9045-848f69e142c9          2022-08  
3  cffb9ed1-1714-11ed-9074-848f69e142cf          2022-08  
4  5bf7d7d7-170d-11ed-9074-848f69e142cf          2022-08

In [29]:
ads_leads.shape

(1212, 17)

Джойним ads_leads и purchases

In [30]:
df = pd.merge(ads_leads, purchases, how='inner', on='client_id')
df.head()

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
0 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
1 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
2 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
3 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
4 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   

  d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  ...  \
0    8404701659      15.0  496.896        2022-08      2022-08-08  ...   
1    8404701661       5.0  124.704        2022-08      2022-08-08  ...   
2    8404701661       5.0  124.704        2022-08      2022-08-08  ...   
3    8404701661       5.0  124.704        2022-08      2022-08-08  ...   
4    8404701665      20.0  385.896        2022-08      2022-08-08  ...   

  d_lead_utm_source d_lead_utm_medium d_lead_utm_campaign d_lead_utm_content  \
0            yandex               cpc            48306450         8404701659   
1            yandex               cpc            48306450         8404701661   
2            yandex               cpc            48306450         8404701661   
3            yandex               cpc            48306450         8404701661   
4            yandex               cpc            48306450         8404701665   

                              client_id leads_year_month purchase_created_at  \
0  6ba575d8-f458-11ec-9070-848f69e142cf          2022-08          2022-06-25   
1  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-03-20   
2  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-03-30   
3  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-08-08   
4  ca65c322-1ac9-11ec-9045-848f69e142c9          2022-08          2022-10-14   

                            purchase_id m_purchase_amount  \
0  1c40ccbb-af39-48d6-8931-e217f665b4fc               0.0   
1  1337d48f-993c-4d58-884d-45126a0f33e4               0.0   
2  327f9ca8-ab45-46cd-ae39-303fce562ab2             285.0   
3  8e090ba8-e323-4189-bb3c-d9e1483a7212            6194.0   
4  cac4fba6-1eb0-4c65-9d9d-10d31ee1d69f             160.0   

   purchases_year_month  
0               2022-06  
1               2022-03  
2               2022-03  
3               2022-08  
4               2022-10  

[5 rows x 21 columns]

In [31]:
df.shape

(1458, 21)

Создаем столбец с кол-вом дней от создания лида до покупки

In [32]:
df['lead_purchase_prd'] = (df['purchase_created_at'] - df['lead_created_at']).dt.days
df['lead_purchase_prd']

0       -44
1      -141
2      -131
3         0
4        67
       ... 
1453     39
1454     41
1455     16
1456      3
1457     11
Name: lead_purchase_prd, Length: 1458, dtype: int64

В переменной min_period держим purchase_id и минимальный период лида до покупки, чтобы соблюсти правило из задания

In [33]:
min_period = df.groupby('purchase_id')['lead_purchase_prd'].min().reset_index().rename(columns={'lead_purchase_prd': 'min_lead_purchase_prd'})

In [34]:
min_period

purchase_id  min_lead_purchase_prd
0     001480ec-fe8d-4ab2-b38b-4c84b8680037                      0
1     0076904d-5f9c-48f5-8793-0f50d7885f89                    -87
2     007de615-97ce-4fca-a06c-20ed6eeed33e                     14
3     00ec1fb8-2d41-418f-b194-6a72f63e6fd6                     57
4     0128dff1-8c42-4baf-9d72-70f1d697f92d                    165
...                                    ...                    ...
1394  fed0b67e-4992-4daa-91f7-b8ba38d4fa3b                     27
1395  fed14c6a-0175-4f32-826d-8220a8a61660                    121
1396  fee74f42-1210-425c-b962-dd219cfc064d                      0
1397  ff82ec66-b8be-41cb-b946-969264e2a8e9                    143
1398  ff9677bd-4b8c-42b4-9ea2-0926903e3e3c                    140

[1399 rows x 2 columns]

Джойним эту переменную с датафреймом

In [35]:
df_2 = pd.merge(df, min_period, how='inner', on='purchase_id')

In [36]:
df_2.head()

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
0 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
1 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
2 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
3 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   
4 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   

  d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  ...  \
0    8404701659      15.0  496.896        2022-08      2022-08-08  ...   
1    8404701661       5.0  124.704        2022-08      2022-08-08  ...   
2    8404701661       5.0  124.704        2022-08      2022-08-08  ...   
3    8404701661       5.0  124.704        2022-08      2022-08-08  ...   
4    8404701665      20.0  385.896        2022-08      2022-08-08  ...   

  d_lead_utm_campaign d_lead_utm_content  \
0            48306450         8404701659   
1            48306450         8404701661   
2            48306450         8404701661   
3            48306450         8404701661   
4            48306450         8404701665   

                              client_id leads_year_month purchase_created_at  \
0  6ba575d8-f458-11ec-9070-848f69e142cf          2022-08          2022-06-25   
1  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-03-20   
2  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-03-30   
3  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-08-08   
4  ca65c322-1ac9-11ec-9045-848f69e142c9          2022-08          2022-10-14   

                            purchase_id m_purchase_amount  \
0  1c40ccbb-af39-48d6-8931-e217f665b4fc               0.0   
1  1337d48f-993c-4d58-884d-45126a0f33e4               0.0   
2  327f9ca8-ab45-46cd-ae39-303fce562ab2             285.0   
3  8e090ba8-e323-4189-bb3c-d9e1483a7212            6194.0   
4  cac4fba6-1eb0-4c65-9d9d-10d31ee1d69f             160.0   

  purchases_year_month lead_purchase_prd  min_lead_purchase_prd  
0              2022-06               -44                    -44  
1              2022-03              -141                   -141  
2              2022-03              -131                   -131  
3              2022-08                 0                      0  
4              2022-10                67                     67  

[5 rows x 23 columns]

Атрибуция лид - покупка:<br>
**-Не должно быть отрицательных периодов**<br>
**-Минимальный период от лида до покупки**<br>
**-Период не больше 15 дней**<br>

In [37]:
df_final = df_2[(df_2['lead_purchase_prd'] >= 0) & (df_2['lead_purchase_prd'] == df_2['min_lead_purchase_prd']) & (df_2['lead_purchase_prd'] <= 15)]
df_final.head(1)

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
3 2022-08-08  xo-for-client-ya       yandex          cpc       48306450   

  d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  ...  \
3    8404701661       5.0  124.704        2022-08      2022-08-08  ...   

  d_lead_utm_campaign d_lead_utm_content  \
3            48306450         8404701661   

                              client_id leads_year_month purchase_created_at  \
3  ddb3e740-a856-11ec-905f-848f69e142cf          2022-08          2022-08-08   

                            purchase_id m_purchase_amount  \
3  8e090ba8-e323-4189-bb3c-d9e1483a7212            6194.0   

  purchases_year_month lead_purchase_prd  min_lead_purchase_prd  
3              2022-08                 0                      0  

[1 rows x 23 columns]

In [38]:
df_final.shape

(495, 23)

Избавляемся от дубликатов

In [39]:
df_final[df_final.duplicated(['client_id', 'lead_id'], keep=False)]

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
59   2022-07-16  xo-for-client-ya       yandex          cpc       48306435   
61   2022-07-16  xo-for-client-ya       yandex          cpc       48306435   
64   2022-07-14  xo-for-client-ya       yandex          cpc       48306450   
68   2022-07-14  xo-for-client-ya       yandex          cpc       48306450   
72   2022-07-13  xo-for-client-ya       yandex          cpc       48306450   
...         ...               ...          ...          ...            ...   
1436 2022-09-06  xo-for-client-ya       yandex          cpc       48306450   
1437 2022-09-06  xo-for-client-ya       yandex          cpc       48306450   
1449 2022-07-07  xo-for-client-ya       yandex          cpc       48306435   
1451 2022-07-07  xo-for-client-ya       yandex          cpc       48306435   
1456 2022-07-07  xo-for-client-ya       yandex          cpc       48306435   

     d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  ...  \
59      8813476161       6.0  119.580        2022-07      2022-07-16  ...   
61      8813476161       6.0  119.580        2022-07      2022-07-16  ...   
64      8404701659      16.0  454.752        2022-07      2022-07-14  ...   
68      8404701659      16.0  454.752        2022-07      2022-07-14  ...   
72      8404701661       9.0  330.828        2022-07      2022-07-13  ...   
...            ...       ...      ...            ...             ...  ...   
1436    8404701659      24.0  983.964        2022-09      2022-09-06  ...   
1437    8404701659      24.0  983.964        2022-09      2022-09-06  ...   
1449    8813476162       3.0   93.108        2022-07      2022-07-07  ...   
1451    8813476162       3.0   93.108        2022-07      2022-07-07  ...   
1456    8813476162       3.0   93.108        2022-07      2022-07-07  ...   

     d_lead_utm_campaign d_lead_utm_content  \
59              48306435         8813476161   
61              48306435         8813476161   
64              48306450         8404701659   
68              48306450         8404701659   
72              48306450         8404701661   
...                  ...                ...   
1436            48306450         8404701659   
1437            48306450         8404701659   
1449            48306435         8813476162   
1451            48306435         8813476162   
1456            48306435         8813476162   

                                 client_id leads_year_month  \
59    14192e61-04e6-11ed-9072-848f69e142cf          2022-07   
61    14192e61-04e6-11ed-9072-848f69e142cf          2022-07   
64    b99c117a-035b-11ed-9072-848f69e142cf          2022-07   
68    b99c117a-035b-11ed-9072-848f69e142cf          2022-07   
72    85b0f39c-0c60-11eb-ac1a-c412f533dba1          2022-07   
...                                    ...              ...   
1436  86a9f03d-2db2-11ed-907b-848f69e142cf          2022-09   
1437  86a9f03d-2db2-11ed-907b-848f69e142cf          2022-09   
1449  a7e5d538-fe8e-11ec-9072-848f69e142cf          2022-07   
1451  a7e5d538-fe8e-11ec-9072-848f69e142cf          2022-07   
1456  a7e5d538-fe8e-11ec-9072-848f69e142cf          2022-07   

     purchase_created_at                           purchase_id  \
59            2022-07-18  8f482919-5e16-4a43-b941-9ec92d148530   
61            2022-07-16  0e837b22-f12a-4402-a326-6c088fab2e51   
64            2022-07-14  57935ae8-01ad-4799-98ab-eb280dcb7a5e   
68            2022-07-15  2cc76849-9ad2-4b4e-9528-066dfc6c513b   
72            2022-07-14  0dbe6a82-ae01-441e-8068-e90898697a65   
...                  ...                                   ...   
1436          2022-09-06  cf6470e3-b23a-40c4-8b24-b8bdd80e9f75   
1437          2022-09-06  7caa9bbd-645d-4297-a4ee-a77b886817b7   
1449          2022-07-10  055dc1b2-f7ce-4e04-b715-7b6711db1e18   
1451          2022-07-10  8158aae8-14a2-4e35-be4d-31225a44b690   
1456          2022-07-10  afdda038-c761-4030-ab00-45b563b660a2   

     m_purchase_amount purch

In [40]:
df_final.drop_duplicates(['client_id', 'lead_id'], inplace=True)

C:\Users\andre\AppData\Local\Temp\ipykernel_6240\419955280.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final.drop_duplicates(['client_id', 'lead_id'], inplace=True)


In [41]:
df_final.shape

(346, 23)

In [42]:
df_final['m_purchase_amount'].value_counts()

m_purchase_amount
0.0        32
290.0      26
5500.0     13
8495.0     11
6600.0     10
           ..
310.0       1
7194.0      1
3850.0      1
10500.0     1
750.0       1
Name: count, Length: 108, dtype: int64

В столбце m_purchase_amount присутствуют покупки на нулевую сумму. От строк с такими данными стоит избавиться

In [43]:
df_final = df_final[df_final['m_purchase_amount'] != 0]

In [44]:
df_final.shape

(314, 23)

Снова проверяем на дубликаты

In [45]:
df_final[df_final.duplicated(['client_id', 'purchase_id'], keep=False)]

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
1225 2022-06-19  xo-for-client-ya       yandex          cpc       48306450   
1226 2022-06-19  xo-for-client-ya       yandex          cpc       48306450   
1229 2022-06-20  xo-for-client-ya       yandex          cpc       48306450   
1230 2022-06-20  xo-for-client-ya       yandex          cpc       48306450   

     d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  ...  \
1225    8404701664      14.0  393.444        2022-06      2022-06-19  ...   
1226    8404701664      14.0  393.444        2022-06      2022-06-19  ...   
1229    8404701664      21.0  362.292        2022-06      2022-06-20  ...   
1230    8404701664      21.0  362.292        2022-06      2022-06-20  ...   

     d_lead_utm_campaign d_lead_utm_content  \
1225            48306450         8404701664   
1226            48306450         8404701664   
1229            48306450         8404701664   
1230            48306450         8404701664   

                                 client_id leads_year_month  \
1225  85457286-1888-11e8-812f-c412f533dba1          2022-06   
1226  85457286-1888-11e8-812f-c412f533dba1          2022-06   
1229  5041b2a0-ce63-11e9-abbc-c412f533dba1          2022-06   
1230  5041b2a0-ce63-11e9-abbc-c412f533dba1          2022-06   

     purchase_created_at                           purchase_id  \
1225          2022-06-20  f9a89ad0-1ac3-4274-9e79-c489b2b1d4b9   
1226          2022-06-20  f9a89ad0-1ac3-4274-9e79-c489b2b1d4b9   
1229          2022-06-20  08dd221c-a3c5-453a-afe8-27feb9189fd6   
1230          2022-06-20  08dd221c-a3c5-453a-afe8-27feb9189fd6   

     m_purchase_amount purchases_year_month lead_purchase_prd  \
1225            7990.0              2022-06                 1   
1226            7990.0              2022-06                 1   
1229            8990.0              2022-06                 0   
1230            8990.0              2022-06                 0   

      min_lead_purchase_prd  
1225                      1  
1226                      1  
1229                      0  
1230                      0  

[4 rows x 23 columns]

In [46]:
df_final.drop_duplicates(['client_id', 'purchase_id'], inplace=True)

In [47]:
df_final.shape

(312, 23)

In [48]:
df_final.sort_values( by='purchase_created_at', inplace=True)

In [49]:
df_final.head()

created_at   d_ad_account_id d_utm_source d_utm_medium d_utm_campaign  \
216 2022-01-07  xo-for-client-ya       yandex          cpc       48306450   
218 2022-01-07  xo-for-client-ya       yandex          cpc       48306450   
215 2022-01-05  xo-for-client-ya       yandex          cpc       48306450   
223 2022-01-09  xo-for-client-ya       yandex          cpc       48306450   
227 2022-01-09  xo-for-client-ya       yandex          cpc       48306450   

    d_utm_content  m_clicks   m_cost ads_year_month lead_created_at  ...  \
216    8404701659       2.0   52.188        2022-01      2022-01-07  ...   
218    8404701664      11.0  216.120        2022-01      2022-01-07  ...   
215    8404701661      11.0  217.380        2022-01      2022-01-05  ...   
223    8404701661      17.0  343.800        2022-01      2022-01-09  ...   
227    8404701661      17.0  343.800        2022-01      2022-01-09  ...   

    d_lead_utm_campaign d_lead_utm_content  \
216            48306450         8404701659   
218            48306450         8404701664   
215            48306450         8404701661   
223            48306450         8404701661   
227            48306450         8404701661   

                                client_id leads_year_month  \
216  f4ea1ca5-d812-11eb-ac42-c412f533dba1          2022-01   
218  6749c1aa-d5a9-11e9-abbe-c412f533dba1          2022-01   
215  7eb80bd0-78bf-11e9-aba2-c412f533dba1          2022-01   
223  576e00f7-712e-11ec-9057-848f69e142cf          2022-01   
227  3cbb1a2f-7129-11ec-9057-848f69e142cf          2022-01   

    purchase_created_at                           purchase_id  \
216          2022-01-07  6efc929e-21d9-4344-b266-7e19a0109aec   
218          2022-01-07  739472b0-1282-49ca-85dd-8b74394fb56f   
215          2022-01-08  41b8d8e1-4489-4afc-a20d-565990f589c9   
223          2022-01-09  8e110fda-b4f5-4796-813e-34941c7d531e   
227          2022-01-09  6f4c408e-d7ca-434d-a42b-dcb6b4e1f96f   

    m_purchase_amount purchases_year_month lead_purchase_prd  \
216             290.0              2022-01                 0   
218             800.0              2022-01                 0   
215             290.0              2022-01                 3   
223            8200.0              2022-01                 0   
227             290.0              2022-01                 0   

     min_lead_purchase_prd  
216                      0  
218                      0  
215                      3  
223                      0  
227                      0  

[5 rows x 23 columns]

Для финальной таблички нам важно знать кол-во лидов на рекламную кампанию

In [50]:
num_leads = ads_leads.groupby(['ads_year_month', 'd_lead_utm_source', 'd_lead_utm_medium', 'd_lead_utm_campaign']).agg({'lead_id': 'nunique'}).reset_index().rename(columns={'lead_id': 'leads_count'})
num_leads.head(10)

ads_year_month d_lead_utm_source d_lead_utm_medium d_lead_utm_campaign  \
0        2022-01            yandex               cpc            48306435   
1        2022-01            yandex               cpc            48306450   
2        2022-02            yandex               cpc            48306435   
3        2022-02            yandex               cpc            48306450   
4        2022-03            yandex               cpc            48306435   
5        2022-03            yandex               cpc            48306450   
6        2022-03            yandex               cpc            48306461   
7        2022-03            yandex               cpc            48306469   
8        2022-03            yandex               cpc            48306473   
9        2022-03            yandex               cpc            48306487   

   leads_count  
0            2  
1           77  
2            3  
3           74  
4           19  
5           34  
6            1  
7            2  
8           22  
9           16

Также нам важно знать кол-во уникальных purchase_id и сумму покупок

In [51]:
sales = df_final.groupby(['ads_year_month', 'd_utm_source', 'd_utm_medium', 'd_utm_campaign']).agg({'purchase_id':'nunique', 'm_purchase_amount':'sum'}).reset_index().rename(columns={'purchase_id': 'sales_num', 
                                                                                                         'm_purchase_amount': 'revenue'})

sales.head(10)

ads_year_month d_utm_source d_utm_medium d_utm_campaign  sales_num   revenue
0        2022-01       yandex          cpc       48306435          1    9400.0
1        2022-01       yandex          cpc       48306450         29  167425.0
2        2022-02       yandex          cpc       48306435          1    9700.0
3        2022-02       yandex          cpc       48306450         22  151144.0
4        2022-03       yandex          cpc       48306435          6   27970.0
5        2022-03       yandex          cpc       48306450         15   81867.0
6        2022-03       yandex          cpc       48306461          1     290.0
7        2022-03       yandex          cpc       48306473          6   28740.0
8        2022-03       yandex          cpc       48306487          3    6980.0
9        2022-03       yandex          cpc       48306494         12   25445.0

Джойним ранее созданные группированные датафреймы

In [52]:
df_grouped = pd.merge(ads_clicks_cost, sales, how='left', on=['ads_year_month', 'd_utm_source', 'd_utm_medium', 'd_utm_campaign'])
df_grouped.head(10)

ads_year_month d_utm_source d_utm_medium d_utm_campaign  clicks  \
0        2022-01       yandex          cpc       48306435    31.0   
1        2022-01       yandex          cpc       48306450   956.0   
2        2022-02       yandex          cpc       48306435    46.0   
3        2022-02       yandex          cpc       48306450   809.0   
4        2022-03       yandex          cpc       48306435   367.0   
5        2022-03       yandex          cpc       48306450   661.0   
6        2022-03       yandex          cpc       48306461    51.0   
7        2022-03       yandex          cpc       48306469    55.0   
8        2022-03       yandex          cpc       48306473   398.0   
9        2022-03       yandex          cpc       48306487   393.0   

   campaign_cost  sales_num   revenue  
0        424.608        1.0    9400.0  
1      17355.324       29.0  167425.0  
2        541.716        1.0    9700.0  
3      16806.492       22.0  151144.0  
4       7587.384        6.0   27970.0  
5      18264.648       15.0   81867.0  
6       3915.288        1.0     290.0  
7       3822.300        NaN       NaN  
8       7383.192        6.0   28740.0  
9       7279.968        3.0    6980.0

In [53]:
df_grouped = pd.merge(df_grouped, num_leads, how='left', left_on=['ads_year_month', 'd_utm_source', 
                                                                   'd_utm_medium', 'd_utm_campaign'], right_on=['ads_year_month', 'd_lead_utm_source', 
                                                                                                                                     'd_lead_utm_medium', 'd_lead_utm_campaign'])


In [54]:
df_grouped.head(10)

ads_year_month d_utm_source d_utm_medium d_utm_campaign  clicks  \
0        2022-01       yandex          cpc       48306435    31.0   
1        2022-01       yandex          cpc       48306450   956.0   
2        2022-02       yandex          cpc       48306435    46.0   
3        2022-02       yandex          cpc       48306450   809.0   
4        2022-03       yandex          cpc       48306435   367.0   
5        2022-03       yandex          cpc       48306450   661.0   
6        2022-03       yandex          cpc       48306461    51.0   
7        2022-03       yandex          cpc       48306469    55.0   
8        2022-03       yandex          cpc       48306473   398.0   
9        2022-03       yandex          cpc       48306487   393.0   

   campaign_cost  sales_num   revenue d_lead_utm_source d_lead_utm_medium  \
0        424.608        1.0    9400.0            yandex               cpc   
1      17355.324       29.0  167425.0            yandex               cpc   
2        541.716        1.0    9700.0            yandex               cpc   
3      16806.492       22.0  151144.0            yandex               cpc   
4       7587.384        6.0   27970.0            yandex               cpc   
5      18264.648       15.0   81867.0            yandex               cpc   
6       3915.288        1.0     290.0            yandex               cpc   
7       3822.300        NaN       NaN            yandex               cpc   
8       7383.192        6.0   28740.0            yandex               cpc   
9       7279.968        3.0    6980.0            yandex               cpc   

  d_lead_utm_campaign  leads_count  
0            48306435          2.0  
1            48306450         77.0  
2            48306435          3.0  
3            48306450         74.0  
4            48306435         19.0  
5            48306450         34.0  
6            48306461          1.0  
7            48306469          2.0  
8            48306473         22.0  
9            48306487         16.0

Считаем CPL и ROAS

In [55]:
df_grouped['CPL'] = (df_grouped['campaign_cost'] / df_grouped['leads_count']).round(2)
df_grouped['ROAS'] = (df_grouped['revenue'] / df_grouped['campaign_cost']).round(2)

df_grouped = df_grouped[['ads_year_month', 'd_utm_source', 'd_utm_medium', 'd_utm_campaign', 'clicks', 'campaign_cost','leads_count', 'revenue','CPL', 'ROAS']]

In [56]:
df_grouped.head()

ads_year_month d_utm_source d_utm_medium d_utm_campaign  clicks  \
0        2022-01       yandex          cpc       48306435    31.0   
1        2022-01       yandex          cpc       48306450   956.0   
2        2022-02       yandex          cpc       48306435    46.0   
3        2022-02       yandex          cpc       48306450   809.0   
4        2022-03       yandex          cpc       48306435   367.0   

   campaign_cost  leads_count   revenue     CPL   ROAS  
0        424.608          2.0    9400.0  212.30  22.14  
1      17355.324         77.0  167425.0  225.39   9.65  
2        541.716          3.0    9700.0  180.57  17.91  
3      16806.492         74.0  151144.0  227.11   8.99  
4       7587.384         19.0   27970.0  399.34   3.69

Убираем строки где стоимость компании равна нулю

In [57]:
df_grouped = df_grouped[df_grouped['campaign_cost'] != 0]

In [58]:
df_grouped

ads_year_month d_utm_source d_utm_medium d_utm_campaign  clicks  \
0         2022-01       yandex          cpc       48306435    31.0   
1         2022-01       yandex          cpc       48306450   956.0   
2         2022-02       yandex          cpc       48306435    46.0   
3         2022-02       yandex          cpc       48306450   809.0   
4         2022-03       yandex          cpc       48306435   367.0   
5         2022-03       yandex          cpc       48306450   661.0   
6         2022-03       yandex          cpc       48306461    51.0   
7         2022-03       yandex          cpc       48306469    55.0   
8         2022-03       yandex          cpc       48306473   398.0   
9         2022-03       yandex          cpc       48306487   393.0   
10        2022-03       yandex          cpc       48306494   453.0   
11        2022-03       yandex          cpc       48306518   318.0   
12        2022-03       yandex          cpc       60279528    67.0   
13        2022-03       yandex          cpc       72000794   205.0   
14        2022-04       yandex          cpc       48306435   355.0   
15        2022-04       yandex          cpc       48306450   552.0   
16        2022-04       yandex          cpc       48306473   517.0   
17        2022-04       yandex          cpc       48306487   500.0   
18        2022-04       yandex          cpc       48306494   462.0   
19        2022-04       yandex          cpc       48306518   148.0   
20        2022-04       yandex          cpc       72000794    25.0   
21        2022-05       yandex          cpc       48306435   300.0   
22        2022-05       yandex          cpc       48306450  1024.0   
23        2022-05       yandex          cpc       48306473   628.0   
24        2022-05       yandex          cpc       48306487   474.0   
25        2022-05       yandex          cpc       48306494   707.0   
26        2022-05       yandex          cpc       48306518   103.0   
27        2022-06       yandex          cpc       48306435   200.0   
28        2022-06       yandex          cpc       48306450  1182.0   
29        2022-06       yandex          cpc       48306473   280.0   
30        2022-06       yandex          cpc       48306487   338.0   
31        2022-06       yandex          cpc       48306494   471.0   
32        2022-06       yandex          cpc       48306518    73.0   
33        2022-07       yandex          cpc       48306435   263.0   
34        2022-07       yandex          cpc       48306450  1212.0   
35        2022-07       yandex          cpc       48306473   293.0   
36        2022-07       yandex          cpc       48306487   269.0   
37        2022-07       yandex          cpc       48306494    42.0   
38        2022-07       yandex          cpc       48306518    47.0   
39        2022-08       yandex          cpc       48306435   337.0   
40        2022-08       yandex          cpc       48306450  1278.0   
41        2022-08       yandex          cpc       48306473    39.0   
42        2022-08       yandex          cpc       48306487   416.0   
43        2022-08       yandex          cpc       48306494   335.0   
45        2022-09       yandex          cpc       48306435   176.0   
46        2022-09       yandex          cpc       48306450   678.0   
47        2022-09       yandex          cpc       48306473    59.0   
48        2022-09       yandex          cpc       48306487   202.0   
49        2022-09       yandex          cpc       48306494     2.0   

    campaign_cost  leads_count   revenue      CPL   ROAS  
0         424.608          2.0    9400.0   212.30  22.14  
1       17355.324         77.0  167425.0   225.39   9.65  
2         541.716          3.0    9700.0   180.57  17.91  
3       16806.492         74.0  151144.0   227.11   8.99  
4        7587.384         19.0   27970.0   399.34   3.69  
5       18264.648         34.0   81867.0   537.20   4.48  
6        3915.288          1.0     290.0  3915.29   0.07  
7        3822.300          2.0

Сохраняем итоговую табличку в excel формате

In [154]:
df_grouped.to_excel('analytics_result.xlsx', index=False)